File: app/src/main/java/it/niedermann/nextcloud/deck/persistence/sync/adapters/db/dao/BoardDao.java
package it.niedermann.nextcloud.deck.persistence.sync.adapters.db.dao;

import androidx.lifecycle.LiveData;
import androidx.room.Dao;
import androidx.room.Query;
import androidx.room.Transaction;

import java.util.List;

import it.niedermann.nextcloud.deck.model.Board;
import it.niedermann.nextcloud.deck.model.full.FullBoard;

@Dao
public interface BoardDao extends GenericDao<Board> {

    @Query("SELECT * FROM board WHERE accountId = :accountId and (deletedAt = 0 or deletedAt is null) and status <> 3 order by title asc")
    LiveData<List<Board>> getBoardsForAccount(final long accountId);

    @Query("SELECT * FROM board WHERE accountId = :accountId and archived = 1 and (deletedAt = 0 or deletedAt is null) and status <> 3 order by title asc")
    LiveData<List<Board>> getArchivedBoardsForAccount(final long accountId);

    @Query("SELECT * FROM board WHERE accountId = :accountId and archived = 0 and (deletedAt = 0 or deletedAt is null) and status <> 3 order by title asc")
    LiveData<List<Board>> getNonArchivedBoardsForAccount(final long accountId);

    @Transaction
    @Query("SELECT * FROM board WHERE accountId = :accountId and archived = :archived and (deletedAt = 0 or deletedAt is null) and status <> 3 order by title asc")
    LiveData<List<FullBoard>> getArchivedFullBoards(long accountId, int archived);

    @Query("SELECT * FROM board WHERE accountId = :accountId and id = :remoteId")
    LiveData<Board> getBoardByRemoteId(final long accountId, final long remoteId);

    @Query("SELECT * FROM board WHERE accountId = :accountId and id = :remoteId")
    Board getBoardByRemoteIdDirectly(long accountId, long remoteId);

    @Query("SELECT * FROM board WHERE localId = :localId")
    Board getBoardByIdDirectly(long localId);

    @Transaction
    @Query("SELECT * FROM board WHERE accountId = :accountId and id = :remoteId")
    FullBoard getFullBoardByRemoteIdDirectly(long accountId, long remoteId);

    @Transaction
    @Query("SELECT * FROM board WHERE accountId = :accountId and localId = :localId")
    FullBoard getFullBoardByLocalIdDirectly(long accountId, long localId);

    @Transaction
    @Query("SELECT * FROM board WHERE accountId = :accountId and (status<>1 or id is null or lastModified <> lastModifiedLocal)")
    List<FullBoard> getLocallyChangedBoardsDirectly(long accountId);

    @Transaction
    @Query("SELECT * FROM board WHERE accountId = :accountId and localId = :localId")
    LiveData<FullBoard> getFullBoardById(final long accountId, final long localId);


    @Query("SELECT b.* FROM board b JOIN stack s ON s.boardId = b.localId JOIN card c ON c.localId = :localCardId")
    Board getBoardByLocalCardIdDirectly(long localCardId);

    @Transaction
    @Query("SELECT b.* FROM board b JOIN stack s ON s.boardId = b.localId JOIN card c ON c.localId = :localCardId and c.stackId = s.localId")
    FullBoard getFullBoardByLocalCardIdDirectly(long localCardId);

    @Transaction
    @Query("SELECT * FROM board WHERE accountId = :accountId")
    List<FullBoard> getAllFullBoards(long accountId);

    @Query("SELECT * FROM board WHERE accountId = :accountId and archived = 0 and permissionEdit = 1 and (deletedAt = 0 or deletedAt is null) and status <> 3 order by title asc")
    LiveData<List<Board>> getBoardsWithEditPermissionsForAccount(long accountId);


    @Query("SELECT b.localId " +
            "FROM card c " +
            "inner join stack s on s.localId = c.stackId " +
            "inner join board b on s.boardId = b.localId " +
            "WHERE c.id = :cardRemoteId and c.accountId =  :accountId")
    LiveData<Long> getLocalBoardIdByCardRemoteIdAndAccountId(long cardRemoteId, long accountId);

    @Query("SELECT count(*) FROM board WHERE accountId = :accountId and archived = 1 and (deletedAt = 0 or deletedAt is null) and status <> 3")
    LiveData<Integer> countArchivedBoards(long accountId);

    @Query("SELECT * FROM board WHERE accountId = :accountId and title = :title")
    Board getBoardForAccountByNameDirectly(long accountId, String title);
}

File: data/local/src/main/java/it/niedermann/nextcloud/deck/data/local/dao/BoardDao.kt
package it.niedermann.nextcloud.deck.data.local.dao

import androidx.room3.Dao
import androidx.room3.Insert
import androidx.room3.OnConflictStrategy
import androidx.room3.Query
import io.reactivex.rxjava3.core.Flowable
import it.niedermann.nextcloud.deck.data.local.entity.BoardEntity
import java.util.concurrent.CompletableFuture

@Dao
interface BoardDao : GenericDao<BoardEntity> {

    @Query("SELECT * FROM Board WHERE accountId = :accountId AND status != 3 ORDER BY archived ASC, title ASC")
    fun getBoardsByAccount(accountId: Long): Flowable<List<BoardEntity>>

    @Query("SELECT * FROM Board WHERE accountId = :accountId")
    fun getAllBoardsByAccount(accountId: Long): CompletableFuture<List<BoardEntity>>

    @Query("SELECT * FROM Board WHERE accountId = :accountId AND remoteId = :remoteId")
    fun getBoardByRemoteId(accountId: Long, remoteId: Long): CompletableFuture<BoardEntity?>

    @Query("SELECT * FROM Board WHERE accountId = :accountId AND status != 1") // 1 = UP_TO_DATE
    fun getChangedBoards(accountId: Long): CompletableFuture<List<BoardEntity>>

    @Query("SELECT * FROM Board WHERE localId = :localId")
    fun getBoardById(localId: Long): CompletableFuture<BoardEntity?>

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    fun insertOrReplace(entity: BoardEntity): CompletableFuture<Long>

    @Query("DELETE FROM Board WHERE localId = :localId")
    fun deleteById(localId: Long): CompletableFuture<Void?>
}
